connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -0,0 +1,537 @@
+        log.debug("Performing task assignment");
+
+        Map<String, ExtendedWorkerState> memberConfigs = new HashMap<>();
+        for (JoinGroupResponseMember member : allMemberMetadata)
+            memberConfigs.put(member.memberId(), IncrementalCooperativeConnectProtocol.deserializeMetadata(ByteBuffer.wrap(member.metadata())));
+
+        // The new config offset is the maximum seen by any member. We always perform assignment using this offset,
+        // even if some members have fallen behind. The config offset used to generate the assignment is included in
+        // the response so members that have fallen behind will not use the assignment until they have caught up.
+        long maxOffset = memberConfigs.values().stream().map(ExtendedWorkerState::offset).max(Long::compare).get();
+        log.debug("Max config offset root: {}, local snapshot config offsets root: {}",
+                  maxOffset, coordinator.configSnapshot().offset());
+
+        Long leaderOffset = ensureLeaderConfig(maxOffset, coordinator);
+        if (leaderOffset == null) {
+            Map<String, ConnectAssignment> assignments = fillAssignments(
+                    memberConfigs.keySet(), Assignment.CONFIG_MISMATCH,
+                    leaderId, memberConfigs.get(leaderId).url(), maxOffset, Collections.emptyMap(),
+                    Collections.emptyMap(), Collections.emptyMap(), 0);
+            return serializeAssignments(assignments);
+        }
+        return performTaskAssignment(leaderId, leaderOffset, memberConfigs, coordinator);
+    }
+
+    private Long ensureLeaderConfig(long maxOffset, WorkerCoordinator coordinator) {
+        // If this leader is behind some other members, we can't do assignment
+        if (coordinator.configSnapshot().offset() < maxOffset) {
+            // We might be able to take a new snapshot to catch up immediately and avoid another round of syncing here.
+            // Alternatively, if this node has already passed the maximum reported by any other member of the group, it
+            // is also safe to use this newer state.
+            ClusterConfigState updatedSnapshot = coordinator.configFreshSnapshot();
+            if (updatedSnapshot.offset() < maxOffset) {
+                log.info("Was selected to perform assignments, but do not have latest config found in sync request. "
+                         + "Returning an empty configuration to trigger re-sync.");
+                return null;
+            } else {
+                coordinator.configSnapshot(updatedSnapshot);
+                return updatedSnapshot.offset();
+            }
+        }
+        return maxOffset;
+    }
+
+    /**
+     * TODO: describe algorithm
+     * @param leaderId
+     * @param maxOffset
+     * @param memberConfigs
+     * @param coordinator
+     * @return
+     */
+    private Map<String, ByteBuffer> performTaskAssignment(String leaderId, long maxOffset,
```

Comment:
How is error handling done for this method? It mutates its own state (previousAssignment) as well as the WorkerCoordinator. If we fail half way through the assignment procedure, how do we ensure a valid overall state?

Reply:
That's a good question. We might need to reset this under certain errors. I'll get back to that.

Reply:
As, mentioned this was an insightful observation. Returning to this after I added several tests on the assignor code. We have the following two cases: 

1) The assignment is computed by the assignor, it contains only revocations and it fails to be delivered. In the next round of rebalancing this failure to change the state of revoked tasks is detected by the assignor, and instead of skipping revocation according to its policy that mandates that there's no subsequent revocations, it re-applies revocation of tasks. This should happen until the assignment containing the revoked tasks succeeds. 
2) The assignment is computed by the assignor, it contains only assignments and it fails to be delivered. Distinguishing this case from the case that active assignments are detected as lost, would result in a more complicated logic of the assignor's state machine. Instead of doing that, the current code, selects to interpret the failed assignment as referring to lost tasks and therefore it enters the deferred rebalancing period, with the specified delay. I believe this keeps things simpler. In a sense whether a worker goes down, or an assignment fails to be delivered, are both considered failures that takes us to the deferred rebalancing logic. 

The current unit tests currently confirm this behavior. WDYT @mumrah ?

Reply:
I think that sounds reasonable. Thanks for the follow-up @kkonstantine!